std/numeric/summation.d

```diff
@@ -0,0 +1,1309 @@
+    assert(sumOfLog2s([-real.nan]).isNaN());
+    assert(sumOfLog2s([real.infinity]) == real.infinity);
+    assert(sumOfLog2s([-real.infinity]).isNaN());
+    assert(sumOfLog2s([ 0.25, 0.25, 0.25, 0.125 ]) == -9);
+}
+
+
+/++
+Computes accurate geometric mean of input range $(D r).
++/
+//A worst-case error of roughly O(ε), independent of N.
+ElementType!Range geometricMean(Range)(Range r)
+    if (isInputRange!Range && isFloatingPoint!(ElementType!Range))
+{
+    static if (hasLength!Range)
+    {
+        immutable n = r.length;
+        immutable s = r.sumOfLog2s();
+    }
+    else
+    {
+        import std.range : tee;
+        size_t n;
+        immutable s = r.tee!((_){n++;}).sumOfLog2s();
+    }
+    return exp2(s / n);
+}
+
+///
+unittest {
+    assert(geometricMean([2.0, 8.0]) == 4.0);
+}
+
+
+/++
+Summation algorithms.
++/
+enum Summation
+{
+    /++
+    Fast summation algorithm.
+    +/
+    fast,
+
+    /++
+    Naive algorithm (one by one).
+    +/
+    naive,
+
+    /++
+    $(WEB en.wikipedia.org/wiki/Pairwise_summation, Pairwise summation) algorithm. Range must be a finite sliceable range.
+    +/
+    pairwise,
+
+    /++
+    $(WEB en.wikipedia.org/wiki/Kahan_summation,
+Kahan summation) algorithm.
+    +/
+    kahan,
+
+    /++
+    $(LUCKY Kahan-Babuška-Neumaier summation algorithm). $(D KBN) gives more accurate results then $(D Kahan).
+    +/
+    kbn,
+
+    /++
+    $(LUCKY Generalized Kahan-Babuška summation algorithm), order 2. $(D KB2) gives more accurate results then $(D Kahan) and $(D KBN).
+    +/
+    kb2,
+
+    /++
+    Precise summation algorithm.
+    The value of the sum is rounded to the nearest representable
+    floating-point number using the $(LUCKY round-half-to-even rule).
+    The result can differ from the exact value on $(D X86), $(D nextDown(proir) <= result &&  result <= nextUp(proir)).
+    The current implementation re-establish special value semantics across iterations (i.e. handling -inf + inf).
+
+    References: $(LINK2 http://www.cs.cmu.edu/afs/cs/project/quake/public/papers/robust-arithmetic.ps,
+        "Adaptive Precision Floating-Point Arithmetic and Fast Robust Geometric Predicates", Jonathan Richard Shewchuk),
+        $(LINK2 http://bugs.python.org/file10357/msum4.py, Mark Dickinson's post at bugs.python.org).
+    +/
+    /+
+    Precise summation function as msum() by Raymond Hettinger in
+    <http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/393090>,
+    enhanced with the exact partials sum and roundoff from Mark
+    Dickinson's post at <http://bugs.python.org/file10357/msum4.py>.
+    See those links for more details, proofs and other references.
+    IEEE 754R floating point semantics are assumed.
+    +/
+    precise,
+
+    /++
+    Performs $(D Pairwise) summation for random access ranges.
+    Otherwise performs $(D KBN) summation of floating point and complex numbers and
+    $(D Kahan) summation of user defined types.
+    +/
+    appropriate,
+}
```

Comment:
You should provide a brief and an extended overview for choosing the most suited algorithms, also explaining their drawbacks. Maybe make a table such as http://dlang.org/phobos/std_algorithm.html.